### DISCUSSION TITLE
Error With PB2NC Tool in MET make_test.log File

### DISCUSSION THREAD

**Discussion by jhicks14** [dtcenter/METplus#2122] [Installation] [answered]
It's me again -- back with some more MET installation issues. I had a previous discussion that involved installation issues that have since been resolved. However, I am trying to now install the newest versions of MET (11.0.2) and METplus. I am able to get all the way to the compilation of MET, to the point where **make test** is run. I am also using a different Conda env than I was last time (/cpc/met/MET11/python-env).

I believe there is an issue with finding the correct version/location of the libgfortran file, which I have had issues with before.

For reference, the version of libgfortran (v5) that I would like installed is here: /cpc/met/MET11/python-env/lib

However, I believe version 5 of libgfortran is being used for compilation (just showing one of the lines of output):

**ldd /cpc/home/jhicks/MET11/MET-11.0.0-beta3/bin/pb2nc**

**libgfortran.so.5 => /cpc/met/MET11/python-env/lib/libgfortran.so.5 (0x00007fa38983e000)**

Here is the error I am getting in the make_test.log file:

**FATAL ERROR (SEGFAULT): Process 37275 got signal 11 @ local time = 2023-04-12 22:38:04Z
FATAL ERROR (SEGFAULT): Look for a core file in /cpc/met/MET11/MET-11.0.2/MET-11.0.2/scripts
FATAL ERROR (SEGFAULT): Process command line: ../src/tools/other/pb2nc/pb2nc ../data/sample_obs/prepbufr/ndas.t00z.prepbufr.tm12.20070401.nr ../out/pb2nc/sampl
e_pb.nc config/PB2NCConfig_G212 -v 2
make[1]: *** [pb2nc] Segmentation fault (core dumped)
make[1]: Leaving directory `/cpc/met/MET11/MET-11.0.2/MET-11.0.2/scripts'
make: *** [test] Error 2**

Here is my compile_MET_all.sh and environmental variable file that I pass in.

[install_met_env.cpc2.txt](https://github.com/dtcenter/METplus/files/11226181/install_met_env.cpc2.txt)
[compile_MET_all.txt](https://github.com/dtcenter/METplus/files/11226182/compile_MET_all.txt)

**Comment by jprestop** [comment]
Hi @jhicks14.  Could you please pass along your config.log file from MET?  I agree that I think the problem lies where you said:

> I believe there is an issue with finding the correct version/location of the libgfortran file, which I have had issues with before.

Do you remember how you solved this problem the last time and can you try that again?

**Reply by jhicks14** [reply]
Hi @jprestop ,

I implemented all of the changes that were suggested last time I had this issue. The one change that seemed to fix the issue before was deleting an unused version of libgfortran from the /lib library of the conda env, although I only see one version in this current env. 

Here is the config.log file from MET:
[config.log](https://github.com/dtcenter/METplus/files/11234990/config.log)

**Reply by jprestop** [reply]
Thank you @jhicks14.  In your compilation script, please save a copy of your existing script and then please try changing:
```
# Update library linker path
export LD_LIBRARY_PATH=${TEST_BASE}/external_libs/lib${MET_PYTHON_LIB:+:$MET_PYTHON_LIB}${MET_NETCDF:+:$MET_NETCDF/lib}${MET_HDF5:+:$MET_HDF5/lib}${MET_BUFRLIB:+:$MET_BUFRLIB}${MET_GRIB2CLIB:+:$MET_GRIB2CLIB}${LIB_JASPER:+:$LIB_JASPER}${LIB_LIBPNG:+:$LIB_LIBPNG}${LIB_Z:+:$LIB_Z}${MET_GSL:+:$MET_GSL/lib}${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}
echo "LD_LIBRARY_PATH = ${LD_LIBRARY_PATH}"
```
to
```
# Update library linker path
export LD_LIBRARY_PATH=${TEST_BASE}/external_libs/lib${MET_NETCDF:+:$MET_NETCDF/lib}${MET_HDF5:+:$MET_HDF5/lib}${MET_BUFRLIB:+:$MET_BUFRLIB}${MET_GRIB2CLIB:+:$MET_GRIB2CLIB}${LIB_JASPER:+:$LIB_JASPER}${LIB_LIBPNG:+:$LIB_LIBPNG}${LIB_Z:+:$LIB_Z}${MET_GSL:+:$MET_GSL/lib}${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}
echo "LD_LIBRARY_PATH = ${LD_LIBRARY_PATH}"
```
Note the removal of ${MET_PYTHON_LIB:+:$MET_PYTHON_LIB}, which may need to be there, but let's give this a try.

and 

```
  export LDFLAGS="${LDFLAGS} -Wl,-rpath,${LIB_DIR}/lib${ADDTL_DIR:+:$ADDTL_DIR/lib}${MET_NETCDF:+:$MET_NETCDF/lib}${MET_HDF5:+:$MET_HDF5/lib}${MET_BUFRLIB:+:$MET_BUFRLIB}${MET_GRIB2CLIB:+:$MET_GRIB2CLIB}${MET_PYTHON:+:$MET_PYTHON/lib}${MET_GSL:+:$MET_GSL/lib}"
 export LDFLAGS="${LDFLAGS} -Wl,-rpath,${LIB_JASPER:+$LIB_JASPER}${LIB_LIBPNG:+:$LIB_PNG}${LIB_Z:+$LIB_Z}"
 export LDFLAGS="${LDFLAGS} ${LIB_JASPER:+-L$LIB_JASPER} ${LIB_LIBPNG:+-L$LIB_LIBPNG} ${MET_HDF5:+-L$MET_HDF5/lib} ${ADDTL_DIR:+-L$ADDTL_DIR/lib}"
```
to
```
  export LDFLAGS="${LDFLAGS} -Wl,-rpath,${LIB_DIR}/lib${ADDTL_DIR:+:$ADDTL_DIR/lib}${MET_NETCDF:+:$MET_NETCDF/lib}${MET_HDF5:+:$MET_HDF5/lib}${MET_BUFRLIB:+:$MET_BUFRLIB}${MET_GRIB2CLIB:+:$MET_GRIB2CLIB}${MET_PYTHON:+:$MET_PYTHON/lib}${MET_GSL:+:$MET_GSL/lib}**${ADDTL_DIR:+:$ADDTL_DIR/lib}**"
 export LDFLAGS="${LDFLAGS} -Wl,-rpath,${LIB_JASPER:+$LIB_JASPER}${LIB_LIBPNG:+:$LIB_PNG}${LIB_Z:+$LIB_Z}"
 export LDFLAGS="${LDFLAGS} ${LIB_JASPER:+-L$LIB_JASPER} ${LIB_LIBPNG:+-L$LIB_LIBPNG} ${MET_HDF5:+-L$MET_HDF5/lib} ${ADDTL_DIR:+-L$ADDTL_DIR/lib}"
```
Note the addition of ${ADDTL_DIR:+:$ADDTL_DIR/lib} at the end of the line.  I know it's already at the beginning.  It won't hurt to leave it there.  I am wondering if it needs to be at the end as well.

Please rerun and let us know how it goes.  If it doesn't fix the problem, please pass along your script, the config.log file and your installation configuration file again.  Thanks!

**Reply by jhicks14** [reply]
Hi @jprestop ,

Unfortunately, it looks like I'm getting the same error after making those changes:

**DEBUG 1: User Config File: config/PB2NCConfig_G212
DEBUG 1: Creating NetCDF File:          ../out/pb2nc/sample_pb.nc
DEBUG 1: Pre-processing BUFR File for variable names from ../data/sample_obs
/prepbufr/ndas.t00z.prepbufr.tm12.20070401.nr
FATAL ERROR (SEGFAULT): Process 14435 got signal 11 @ local time = 2023-04-1
4 19:40:02Z
FATAL ERROR (SEGFAULT): Look for a core file in /cpc/met/MET11/MET-11.0.2/ME
T-11.0.2/scripts
FATAL ERROR (SEGFAULT): Process command line: ../src/tools/other/pb2nc/pb2nc
 ../data/sample_obs/prepbufr/ndas.t00z.prepbufr.tm12.20070401.nr ../out/pb2n
c/sample_pb.nc config/PB2NCConfig_G212 -v 2
make[1]: *** [pb2nc] Segmentation fault (core dumped)
make[1]: Leaving directory `/cpc/met/MET11/MET-11.0.2/MET-11.0.2/scripts'
make: *** [test] Error 2**

[config.log](https://github.com/dtcenter/METplus/files/11236548/config.log)
[compile_MET_all2.txt](https://github.com/dtcenter/METplus/files/11236549/compile_MET_all2.txt)
[install_met_env.cpc2.txt](https://github.com/dtcenter/METplus/files/11236550/install_met_env.cpc2.txt)

**Reply by jprestop** [reply]
Thanks for trying, @jhicks14.  Ok, let's try something else. Please go ahead and switch back to the original compiile_MET_all.sh script that you were using.  I have attached an install_met_env.cpc2.attempt1.txt file where I have added the following:
```
export FCLIBS="-L${ADDTL_DIR}/lib -lgfortran"
export FLIBS="-L${ADDTL_DIR}/lib -lgfortran"
```
Please try running with this new file and let us know how it goes. If it doesn't fix the problem, please pass along your script, the config.log file and your installation configuration file again. Thanks!
[install_met_env.cpc2.attempt1.txt](https://github.com/dtcenter/METplus/files/11252602/install_met_env.cpc2.attempt1.txt)

**Reply by jhicks14** [reply]
Hi @jprestop 

I just got a chance to try out that install file for the environmental vars, and it looks like I'm still running into the same issues.

**DEBUG 1: Pre-processing BUFR File for variable names from ../data/sample_obs/prepbufr
/ndas.t00z.prepbufr.tm12.20070401.nr
FATAL ERROR (SEGFAULT): Process 16314 got signal 11 @ local time = 2023-04-17 20:10:0
1Z
FATAL ERROR (SEGFAULT): Look for a core file in /cpc/met/MET11/MET-11.0.2/MET-11.0.2/
scripts
FATAL ERROR (SEGFAULT): Process command line: ../src/tools/other/pb2nc/pb2nc ../data/
sample_obs/prepbufr/ndas.t00z.prepbufr.tm12.20070401.nr ../out/pb2nc/sample_pb.nc con
fig/PB2NCConfig_G212 -v 2
make[1]: *** [pb2nc] Segmentation fault (core dumped)
make[1]: Leaving directory `/cpc/met/MET11/MET-11.0.2/MET-11.0.2/scripts'
make: *** [test] Error 2**

[compile_MET_all.txt](https://github.com/dtcenter/METplus/files/11255402/compile_MET_all.txt)
[install_met_env.cpc2.attempt1.txt](https://github.com/dtcenter/METplus/files/11255403/install_met_env.cpc2.attempt1.txt)
[config.log](https://github.com/dtcenter/METplus/files/11255404/config.log)

**Reply by jprestop** [reply]
Hi @jhicks14.  What I was intending to test with that change, was not actually tested.  Based on [this page](https://www.gnu.org/software/autoconf/manual/autoconf-2.60/html_node/Fortran-Compiler.html), I thought that the variable I wanted was FCLIBS (all one word with no underscore), however, I still see only FC_LIBS in your config file.  I'm not sure why exactly that is, but let's have you try changing the following in your install_met_env.cpc2.attempt1.txt file.  Please change:
```
export FCLIBS="-L${ADDTL_DIR}/lib -lgfortran"
```
to 
```
export FC_LIBS="-L${ADDTL_DIR}/lib -lgfortran"
```
and try rerunning.

Please try running with this change and let us know how it goes. If it doesn't fix the problem, please pass along your script, the config.log file and your installation configuration file again. Thanks!

**Reply by jprestop** [reply]
@jhicks14 Also, could you please run the following and let me know the output:
```
ls /usr/lib64/ | grep gfortran
```

**Reply by jhicks14** [reply]
Hi @jprestop 

Looks like I'm still getting the same error. This is the output from **ls /usr/lib64/ | grep gfortran**:

libgfortran.so.1
libgfortran.so.1.0.0
libgfortran.so.3
libgfortran.so.3.0.0

[compile_MET_all.txt](https://github.com/dtcenter/METplus/files/11266699/compile_MET_all.txt)
[install_met_env.cpc2.attempt1.txt](https://github.com/dtcenter/METplus/files/11266700/install_met_env.cpc2.attempt1.txt)
[config.log](https://github.com/dtcenter/METplus/files/11266701/config.log)

**Reply by jprestop** [reply]
Hi @jhicks14.  I just wanted to check in to be sure you ran the compile script with the install_met_env.cpc2.attempt1.txt file?  I ask because the config.log file does not show the -L${ADDTL_DIR}/lib added to FC_LIBS, which we included in the install_met_env.cpc2.attempt1.txt file:
```
export FC_LIBS="-L${ADDTL_DIR}/lib -lgfortran"
export FLIBS="-L${ADDTL_DIR}/lib -lgfortran"
```
From config.log:
```
FC_LIBS='-lgfortran'
FLIBS='-L/usr/lib64/lib -lgfortran'
```

**Reply by jhicks14** [reply]
Hi @jprestop 

I'm not sure what the issue is now. We just upgraded to the latest version of RedHat Linux, so this error could be due to that, but basically it no longer gets to the "make test" part of the install. The config.log file looks fine, but the error lies in the make.log file for MET. I ran this using the **install_met_env.cpc2.attempt1.txt** file.

Should I just re-run all of the prerequisite programs (e.g. comment out the bottom part of the install file that contains the env vars)?


[compile_MET_all.txt](https://github.com/dtcenter/METplus/files/11299426/compile_MET_all.txt)
[install_met_env.cpc2.attempt1.txt](https://github.com/dtcenter/METplus/files/11299427/install_met_env.cpc2.attempt1.txt)
[config.log](https://github.com/dtcenter/METplus/files/11299428/config.log)
[make.log](https://github.com/dtcenter/METplus/files/11299429/make.log)

**Reply by jprestop** [reply, marked as answer]
Hi @jhicks14. My apologies for my delayed response.  I was out of the office.  Upgrading to the latest version of RedHat Linux, could have caused problems.  It's hard to say.  Since you weren't experiencing these problems previously and I don't see anything obvious in the files you sent, please go ahead an reinstall all of the library dependencies as you suggested, by commenting out the bottom part of the env vars file (starting with MET_GSL and ending with MET_NETCDF)  Also, please leave this line commented the whole time and see what happens. 
```
export LD_LIBRARY_PATH=/cpc/met/MET11/$PYTHON_VER/lib
```
Please let us know how it goes.

**Reply by jhicks14** [reply]
Hi @jprestop

I reinstalled all of the library dependencies (i.e., commented out all of the bottom part of the env vars file), kept the line you mentioned commented out, and received an error compiling GSL (couldn't be found in /usr/lib64). Fortunately, I fixed this by commenting out this line:

**export LDFLAGS=/usr/lib64:${LDFLAGS}** 

Then everything ran successfully with no errors in the make_test.log file!

Thanks again for the help!

**Reply by jprestop** [reply]
@jhicks14 Excellent!  Thank you so much for letting us know.  I'll go ahead and mark this as answered, close, and lock it to prevent future posts in this thread.
